File: packages/faasit-runtime/javascript/src/runtime/AliyunRuntime.ts
import { BaseFaasitRuntime, CallResult, FaasitRuntime, FaasitRuntimeMetadata, InputType, StorageMethods } from "./FaasitRuntime";
import FC_Open20210406, * as $FC_Open20210406 from '@alicloud/fc-open20210406';
import * as $OpenApi from '@alicloud/openapi-client';
import Util, * as $Util from '@alicloud/tea-util';
import OSS from 'ali-oss';

function newOSSBucketClient() {
    return new OSS({
        region: process.env.ALIBABA_CLOUD_OSS_REGION, // oss-cn-hangzhou
        accessKeyId: process.env.ALIBABA_CLOUD_ACCESS_KEY_ID || '',
        accessKeySecret: process.env.ALIBABA_CLOUD_ACCESS_KEY_SECRET || '',
        bucket: process.env.ALIBABA_CLOUD_OSS_BUCKET_NAME // faasit
    })
}

function helperInvokeAliyunFunction(fnName: string, event: any) {
    const config = new $OpenApi.Config({
        accessKeyId: process.env.ALIBABA_CLOUD_ACCESS_KEY_ID,
        accessKeySecret: process.env.ALIBABA_CLOUD_ACCESS_KEY_SECRET,
        endpoint: `${process.env.ALIBABA_CLOUD_PRODUCT_CODE}.${process.env.ALIBABA_CLOUD_REGION}.fc.aliyuncs.com`
    });
    const client = new FC_Open20210406(config)
    const invokeFunctionRequests = new $FC_Open20210406.InvokeFunctionRequest({
        body: event ? Util.toBytes(JSON.stringify(event)) : ''
    })

    return client.invokeFunction(`${process.env.ALIBABA_CLOUD_SERVICE}`, fnName, invokeFunctionRequests)
}

// @see https://help.aliyun.com/zh/fc/http-handler
export interface AliyunHttpRuntimeOptions {
    req: { body: Buffer }
    resp: {
        setHeader: (key: string, value: string) => void
        send: (body: string) => void
    }
    context: any,
    metadata: FaasitRuntimeMetadata
}

export class AliyunHttpRuntime extends BaseFaasitRuntime {
    name: string = "aliyun";
    bucket: OSS;

    constructor(private opt: AliyunHttpRuntimeOptions) {
        super()
        this.bucket = newOSSBucketClient();
    }

    metadata(): FaasitRuntimeMetadata {
        return this.opt.metadata
    }

    input(): InputType {
        return JSON.parse(this.opt.req.body.toString())
    }

    output(returnObject: any) {
        this.opt.resp.setHeader('content-type', 'application/json')
        this.opt.resp.send(JSON.stringify(returnObject))
        return returnObject
    }

    async call(fnName: string, fnParams?: {
        sequence?: number;
        input: object;
    }): Promise<CallResult> {
        const res = await helperInvokeAliyunFunction(fnName, fnParams?.input);
        const result = res.body.toString();
        return { output: JSON.parse(result) };
    }

    storage: StorageMethods = {
        put: async (filename: string, data: Buffer): Promise<void> => {
            try {
                const result = await this.bucket.put(filename, data)
                if (result.res.status >= 200 && result.res.status < 300) {
                    console.log(`[Info] Put data into ${filename} successfully.`);
                } else {
                    console.log(result);
                }
            } catch (error) {
                console.log(error);
            }
        },

        get: async (filename: string, timeout = -1): Promise<Buffer | null> => {
            const start_t = Date.now();
            const tryGetObject = async (): Promise<Buffer | null> => {
                if (await this.storage.exists(filename)) {
                    const result = await this.bucket.get(filename);
                    return result.content as Buffer;
                } else if (timeout > 0 && Date.now() - start_t > timeout) {
                    return null;
                } else {
                    return tryGetObject();
                }
            };

            return tryGetObject();
        },

        list: async (): Promise<string[]> => {
            const result = await this.bucket.list({
                'max-keys': 1000
            }, {
                timeout: 3000
            });
            return result.objects.map(object => object.name);
        },

        exists: async (filename: string): Promise<boolean> => {
            try {
                await this.bucket.head(filename);
                return true;
            } catch (error) {
                if (error.code === 'NoSuchKey') {
                    return false;
                }
            }
            return false;
        },

        del: async (filename: string): Promise<void> => {
            try {
                const result = await this.bucket.delete(filename);
                if (result.res.status >= 200 && result.res.status < 300) {
                    console.log(`[Info] Delete ${filename} successfully.`);
                } else {
                    console.log(result);
                }
            } catch (error) {
                console.log(error);
            }
        }
    };
}

export class AliyunEventRuntime extends BaseFaasitRuntime {
    name: string = "aliyun";
    bucket: OSS;

    constructor(private opt: { event: Buffer, context: any, callback: (error: any, data: object) => void, metadata: FaasitRuntimeMetadata }) {
        super()
        this.bucket = newOSSBucketClient();
    }

    metadata(): FaasitRuntimeMetadata {
        return this.opt.metadata
    }

    async call(fnName: string, fnParams?: {
        sequence?: number;
        input: object;
    }): Promise<CallResult> {
        const res = await helperInvokeAliyunFunction(fnName, fnParams?.input);
        const result = res.body.toString();
        return { output: JSON.parse(result) };
    }

    input() {
        return JSON.parse(this.opt.event.toString())
    }

    output(returnObject: any) {
        this.opt.callback(null, returnObject)
        return returnObject
    }

    storage: StorageMethods = {
        put: async (filename: string, data: Buffer): Promise<void> => {
            try {
                const result = await this.bucket.put(filename, data)
                if (result.res.status >= 200 && result.res.status < 300) {
                    console.log(`[Info] Put data into ${filename} successfully.`);
                } else {
                    console.log(result);
                }
            } catch (error) {
                console.log(error);
            }
        },

        get: async (filename: string, timeout = -1): Promise<Buffer | null> => {
            const start_t = Date.now();
            const tryGetObject = async (): Promise<Buffer | null> => {
                if (await this.storage.exists(filename)) {
                    const result = await this.bucket.get(filename);
                    return result.content as Buffer;
                } else if (timeout > 0 && Date.now() - start_t > timeout) {
                    return null;
                } else {
                    return tryGetObject();
                }
            };

            return tryGetObject();
        },

        list: async (): Promise<string[]> => {
            const result = await this.bucket.list({
                'max-keys': 1000
            }, {
                timeout: 3000
            });
            return result.objects.map(object => object.name);
        },

        exists: async (filename: string): Promise<boolean> => {
            try {
                await this.bucket.head(filename);
                return true;
            } catch (error) {
                if (error.code === 'NoSuchKey') {
                    return false;
                }
            }
            return false;
        },

        del: async (filename: string): Promise<void> => {
            try {
                const result = await this.bucket.delete(filename);
                if (result.res.status >= 200 && result.res.status < 300) {
                    console.log(`[Info] Delete ${filename} successfully.`);
                } else {
                    console.log(result);
                }
            } catch (error) {
                console.log(error);
            }
        }
    };
}

File: packages/faasit-runtime/python/faasit_runtime/runtime/aliyun_runtime.py
import time
from faasit_runtime.runtime import FaasitRuntime
from alibabacloud_fc_open20210406.client import Client as FC_Open20210406Client
from alibabacloud_tea_openapi import models as open_api_models
from alibabacloud_fc_open20210406 import models as fc__open20210406_models
from alibabacloud_tea_util import models as util_models
from alibabacloud_tea_util.client import Client as UtilClient
from typing import Any, List
import os
import json
from dotenv import load_dotenv,find_dotenv
from faasit_runtime.runtime.faasit_runtime import StorageMethods
import oss2

# 获取用户进程中的环境变量文件
load_dotenv(find_dotenv(usecwd=True))
def helper_invoke_aliyun_function(fnName: str, event: Any):
    config = open_api_models.Config(
        access_key_id=os.environ['FAASIT_SECRET_ALIYUN_ACCESS_KEY_ID'],
        access_key_secret=os.environ['FAASIT_SECRET_ALIYUN_ACCESS_KEY_SECRET']
    )
    config.endpoint = f"{os.environ['FAASIT_SECRET_ALIYUN_ACCOUNT_ID']}.{os.environ['FAASIT_SECRET_ALIYUN_REGION']}.fc.aliyuncs.com"
    
    client = FC_Open20210406Client(config)
    invoke_func_req = fc__open20210406_models.InvokeFunctionRequest(
        body=UtilClient.to_bytes(event or {}),
    )
    return client.invoke_function('faasit', fnName, invoke_func_req)

class AliyunRuntime(FaasitRuntime):
    name: str = 'aliyun'
    def __init__(self, arg0, arg1) -> None:
        super().__init__()
        self.event = arg0
        self.context = arg1
        self._storage = self.AliyunStorage()

    def input(self):
        return json.loads(self.event)

    def output(self, data):
        return data

    def call(self, fn_name:str, event: Any):
        result = helper_invoke_aliyun_function(fn_name, event)
        result = str(result.body)
        return json.loads(result)

    async def tell(self):
        pass

    class AliyunStorage(StorageMethods):
        def __init__(self):
            auth = oss2.Auth(os.environ['ALIBABA_CLOUD_ACCESS_KEY_ID'], 
                            os.environ['ALIBABA_CLOUD_ACCESS_KEY_SECRET'])
            endpoint = f'https://{os.environ["ALIBABA_CLOUD_OSS_REGION"]}.aliyuncs.com'
            self.bucket = oss2.Bucket(auth, 
                            endpoint, 
                            os.environ['ALIBABA_CLOUD_OSS_BUCKET_NAME'])

        def put(self, filename, data: bytes) -> None:
            res = self.bucket.put_object(filename, data)
            if 200 <= res.resp.status < 300:
                print(f"[storage put] Put data into {filename} successfully.")
            else:
                print(res)

        def get(self, filename, timeout = -1) -> bytes:
            start_t = time.time()
            while not self.bucket.object_exists(filename):
                time.sleep(0.001)
                if timeout > 0:
                    if time.time() - start_t > timeout / 1000: return None
            return self.bucket.get_object(filename).read()

        def list(self) -> List:
            return [sbj.key for sbj in self.bucket.list_objects_v2().object_list]

        def exists(self, filename: str) -> bool:
            return self.bucket.object_exists(filename)

        def delete(self, filename: str) -> None:
            if self.bucket.object_exists(filename):    
                self.bucket.delete_object(filename)
                print(f"[storage delete] Delete {filename} successfully.")
            else:
                print(f"[storage delete] {filename} is not exist.")
                